### PR TITLE
[STRAT-3865] | Re-authentication Flow for CreateAudience and GetAudience to handle 401

### PR DESCRIPTION
**Pull request by Innovative-GauravKochar**
This Pull request is to make re-authentication flow for createAudience and getAudience whenever it throws 401 error.
Jira ticket :- https://segment.atlassian.net/jira/software/c/projects/STRATCONN/boards/310?assignee=63617339fc0cc7a600b03c6b&selectedIssue=STRATCONN-3865

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

Tested completed Successfully in [Staging](https://docs.google.com/document/d/1cfCFtbaO8YX5Y-VyVKp1JhpJWVUdN4O5JOLbW4HX7i4/edit#heading=h.x9dgiyfuowsl)

### CODE REVIEW

File: packages/core/src/destination-kit/index.ts [hunk 1]
```diff
@@ -422,41 +422,94 @@ export class Destination<Settings = JSONObject, AudienceSettings = JSONObject> {
...
+      }
+      const opts = this.extendRequest?.(context) ?? {}
+      const requestClient = createRequestClient({ ...opts, statsContext: context.statsContext })
+      return await audienceConfig?.getAudience(requestClient, getAudienceInput)
     }
-    const options = this.extendRequest?.(context) ?? {}
-    const requestClient = createRequestClient({ ...options, statsContext: context.statsContext })
 
-    return audienceDefinition.audienceConfig?.getAudience(requestClient, getAudienceInput)
+    const onFailedAttempt = async (error: ResponseError & HTTPError) => {
```
Comment (varadarajan-tw) [inline packages/core/src/destination-kit/index.ts:489]: Could you try and see if we can simplify this block of code further? We seem to repeat the same across all the execute blocks?

File: packages/core/src/destination-kit/index.ts [hunk 2]
```diff
@@ -422,41 +422,94 @@ export class Destination<Settings = JSONObject, AudienceSettings = JSONObject> {
...
+      }
+      const opts = this.extendRequest?.(context) ?? {}
+      const requestClient = createRequestClient({ ...opts, statsContext: context.statsContext })
+      return await audienceConfig?.createAudience(requestClient, createAudienceInput)
     }
-    const options = this.extendRequest?.(context) ?? {}
-    const requestClient = createRequestClient({ ...options, statsContext: context.statsContext })
 
-    return audienceDefinition.audienceConfig?.createAudience(requestClient, createAudienceInput)
+    const onFailedAttempt = async (error: ResponseError & HTTPError) => {
```
Comment (varadarajan-tw) [inline packages/core/src/destination-kit/index.ts:444]: Could you add tests to validate this retry and the new flow?

File: packages/core/src/destination-kit/index.ts [hunk 3]
```diff
@@ -857,4 +824,31 @@ export class Destination<Settings = JSONObject, AudienceSettings = JSONObject> {
     const { subcription, subscriptions, oauth, ...otherSettings } = settings
     return otherSettings as unknown as Settings
   }
+  // Refreshes the token and update it in setting in case of 401(Unauthorized).
+  async runFailedAttempt(error: ResponseError & HTTPError, settings: JSONObject, options?: OnEventOptions) {
```
Comment (varadarajan-tw) [inline packages/core/src/destination-kit/index.ts:828]: should rename this method to indicate the behaviour like the comment suggests? It doesn't just run failed attempt right?

File: packages/core/src/__tests__/destination-kit.test.ts [hunk 4]
```diff
@@ -77,6 +79,70 @@ const destinationOAuth2: DestinationDefinition<JSONObject> = {
...
+    },
+    // Mocked createAudience Handler
+    async createAudience(_request, createAudienceInput) {
+      const settings: any = createAudienceInput.settings
+
+      if (settings.oauth.access_token == 'fresh-token' || settings.oauth.access_token == 'valid-access-token') {
+        return new Promise((resolve, _reject) => {
+          setTimeout(() => {
+            resolve({ externalId: '123456789' })
+          }, 3)
```
Comment (varadarajan-tw) [inline packages/core/src/__tests__/destination-kit.test.ts:114]: why timeout?

File: packages/core/src/__tests__/destination-kit.test.ts [hunk 5]
```diff
@@ -953,4 +1019,94 @@ describe('destination kit', () => {
...
+          }
+        }
+        const destinationTest = new Destination(audienceDestination)
+        const spy = jest.spyOn(authentication, 'refreshAccessToken')
+        const res = await destinationTest.getAudience(createAudienceInput)
+        expect(res).toEqual({ externalId: '366170701270726115' })
+        expect(spy).toHaveBeenCalledTimes(1)
+      })
+
+      test('Will not refresh access-token if token is already valid', async () => {
```
Comment (varadarajan-tw) [inline packages/core/src/__tests__/destination-kit.test.ts:1091]: should there be another test case to ensure that for non-401 error, the block is not retried?

Comment (varadarajan-tw) [reply packages/core/src/__tests__/destination-kit.test.ts:1091]: there should also be another test case for non-oauth auth scheme

**Review by varadarajan-tw — COMMENTED** [review]
Nice work! Left few minor comments around test coverage.

**Comment by Innovative-GauravKochar** [comment]
Thanks @varadarajan-tw for your valuable inputs. I have updated the PR , Please have a look!

**Review by varadarajan-tw — APPROVED** [review]

**Comment by joe-ayoub-segment** [comment]
hi @Innovative-GauravKochar PR deployed

**Pull request merged**